### PR TITLE
Don't call :finish hook if there was an exception

### PR DESCRIPTION
**Pull request by ajaska**
I couldn't find any good reason for the :finish hook to be inside an ensure block.

But because it is, the :finish hook will be called with a `nil` value for `result` if there is an exception on the :start hook or `yield` call. This is (IMHO) unexpected behavior, and I was quite confused as to why my function (which should always be called with a non-nil value) was receiving `nil`.

### CODE REVIEW

**Comment by grosser** [comment]
I think this is a feature without tests ... it is supposed to work when using `raise Parallel::Break` ... can you add a test to see that this still works ?

**Comment by ajaska** [comment]
Hmm, I'm not sure I understand. When you `raise Parallel::Break`, do you still want the finish hook to be called for that worker? From what I can tell, I'm not sure that that would be useful as

- You won't have a usable result from that worker
- You are terminating early, so progress information isn't useful

I suppose it might be worth doing for compatibility reasons?

**Comment by ajaska** [comment]
I went ahead and pushed up test cases for my interpretation, let me know if you think `Parallel::Break` should be a special case, and I'd be happy to push up that version.

File: spec/cases/with_exception_in_start_before_finish.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+require './spec/cases/helper'
+
+method = ENV.fetch('METHOD')
+in_worker_type = "in_#{ENV.fetch('WORKER_TYPE')}".to_sym
+
+begin
+  start = lambda do |_item, _index|
```
Comment (grosser) [inline spec/cases/with_exception_in_start_before_finish.rb]: _xx means it's unused ... but here it is used :)

File: spec/cases/with_break_before_finish.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  Parallel.public_send(method, 1..10, in_worker_type => 4, finish: finish) do |x|
+    sleep 0.2
+    if x != 3
+      raise Parallel::Break
+    end
+    print x
+    x
+  end
+rescue
```
Comment (grosser) [inline spec/cases/with_break_before_finish.rb]: the rescue should not be necessary

File: spec/cases/with_exception_before_finish.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  Parallel.public_send(method, 1..10, in_worker_type => 4, finish: finish) do |x|
+    if x != 3
+      sleep 0.2
+      raise 'foo'
+    end
+    print x
+    x
+  end
+rescue
```
Comment (grosser) [inline spec/cases/with_exception_before_finish.rb]: maybe also raise something specific and then rescue only that ... to make sure we are not raising random errors

File: spec/cases/with_break_before_finish.rb (hunk 2 above)
Comment (grosser) [inline spec/cases/with_break_before_finish.rb]: the expected result would be finish getting called with 1,2,3 since these items finished ?

Comment (ajaska) [inline spec/cases/with_break_before_finish.rb]: ah, you're right, I structured this testcase poorly

**Comment by ajaska** [comment]
Thanks for the feedback, I went ahead and made those changes. Let me know about other changes/if you want me to rebase or something.

**Comment by grosser** [comment]
nice tests! :)

**Comment by grosser** [comment]
1.6.2 is out!

**Pull request merged**